plugins/src/gcc/imp.rs

```diff
@@ -0,0 +1,1274 @@
+    /// calculation of added value according to gcc section 5.5
+    last_increase_on_delay: Option<time::Instant>,
+    last_decrease_on_delay: time::Instant,
+
+    /// Bitrate target based on loss for all video streams.
+    target_bitrate_on_loss: Bitrate,
+
+    last_increase_on_loss: time::Instant,
+    last_decrease_on_loss: time::Instant,
+
+    /// Exponential moving average, updated when bitrate is
+    /// decreased
+    ema: ExponentialMovingAverage,
+
+    last_control_op: BandwidthEstimationOp,
+
+    min_bitrate: Bitrate,
+    max_bitrate: Bitrate,
+
+    detector: Detector,
+
+    periodic_push: Option<gst::PeriodicClockId>,
+
+    // Implemented like a leaky bucket
+    buffers: VecDeque<gst::Buffer>,
+
+    flow_return: Result<gst::FlowSuccess, gst::FlowError>,
+    last_push: time::Instant,
+}
+
+impl Default for State {
+    fn default() -> Self {
+        Self {
+            target_bitrate_on_delay: DEFAULT_ESTIMATED_BITRATE,
+            target_bitrate_on_loss: DEFAULT_ESTIMATED_BITRATE,
+            last_increase_on_loss: time::Instant::now(),
+            last_decrease_on_loss: time::Instant::now(),
+            ema: Default::default(),
+            last_increase_on_delay: None,
+            last_decrease_on_delay: time::Instant::now(),
+            min_bitrate: DEFAULT_MIN_BITRATE,
+            max_bitrate: DEFAULT_MAX_BITRATE,
+            detector: Detector::new(),
+            buffers: Default::default(),
+            estimated_bitrate: DEFAULT_ESTIMATED_BITRATE,
+            last_control_op: BandwidthEstimationOp::Increase("Initial increase".into()),
+            flow_return: Err(gst::FlowError::Flushing),
+            periodic_push: None,
+            last_push: time::Instant::now(),
+        }
+    }
+}
+
+impl State {
+    fn create_buffer_list(&mut self, bwe: &super::BandwidthEstimator) -> gst::BufferList {
+        let now = time::Instant::now();
+        let elapsed = Duration::from_std(now - self.last_push).unwrap();
+        let mut budget = (elapsed.num_nanoseconds().unwrap())
+            .mul_div_round(
+                self.estimated_bitrate as i64,
+                gst::ClockTime::SECOND.nseconds() as i64,
+            )
+            .unwrap();
+        let total_budget = budget;
+        let mut remaining = self.buffers.iter().map(|b| b.size() as f64).sum::<f64>() * 8.;
+        let total_size = remaining;
+
+        let mut list = gst::BufferList::new();
+        let mutlist = list.get_mut().unwrap();
+
+        // Leak the bucket so it can hold at most 30ms of data
+        let maximum_remaining_bits = 30. * self.estimated_bitrate as f64 / 1000.;
+        while (budget > 0 || remaining > maximum_remaining_bits) && !self.buffers.is_empty() {
+            let buf = self.buffers.pop_back().unwrap();
+            let n_bits = buf.size() * 8;
+
+            mutlist.add(buf);
+            budget = budget.saturating_sub(n_bits as i64);
+            remaining -= n_bits as f64;
```

Comment:
We also want to perform stuffing here right?

Reply:
Yes

Reply:
Yes